src/Integrator/BDHI/DoublyPeriodic/DPStokesSlab.cuh

```diff
@@ -96,6 +96,44 @@ namespace uammd{
 	return {particleVelocities, particleAngularVelocities};
       }
 
+      // compute average velocity in the x direction as a function of z
+      template<class PosIterator, class ForceIterator>
+      std::vector<double>
+      computeAverageVelocity(PosIterator pos, ForceIterator forces,
+			     int numberParticles, cudaStream_t st = 0){
+      	cudaDeviceSynchronize();
+      	System::log<System::DEBUG2>("[DPStokes] Computing displacements");
+      	auto gridData = ibm->spreadForces(pos, forces, numberParticles, st);
+      	auto gridForceCheb = fct->forwardTransform(gridData, st);
+      	FluidData<complex> fluid = solveBVPVelocity(gridForceCheb, st);
+      	if(mode != WallMode::none){
+      	  correction->correctSolution(fluid, gridForceCheb, st);
+      	}
+	int nx = this->grid.cellDim.x, ny = this->grid.cellDim.y, nz = this->grid.cellDim.z;
+	// std::cout << "printing average velocity" << std::endl;
+	auto xvel = fluid.velocity.m_x;
```

Comment:
With a small change you can make this function average in either X or Y:
```c++ 
std::vector<complex> chebCoeff;
if(direction == 0) chebCoeff = getZeroModeChebCoeff(fluid.velocity.m_x, grid.cellDim);
else if(direction == 1)  chebCoeff = getZeroModeChebCoeff(fluid.velocity.m_y, grid.cellDim);
else throw std::runtime_error("[DPStokesSlab] Can only average in direction X (0) or Y (1)");
```
Where the getZeroMode function here you can write at the start of the file, inside the dpstokes namespace as:
```c++
namespace detail{

template<class ComplexContainer>
auto getZeroModeChebCoeff(const ComplexContainer &fourierChebGridData, int3 n){
    std::vector<complex> chebCoeff(n.z);
    for(int i=0;i<nz;i++){
	  chebCoeff[i] = fourierChebGridData[(nx/2+1)*ny*i]/(nx*ny);
    }
   return chebCoeff;
} 
}
```

Reply:
Thanks. Implemented in the code. I didn't use the new namesapce detail though.